### PR TITLE
Read data from SSL Game Controller

### LINKED ISSUE #72: Read data from SSL-Refbox
In addition to SSL vision data that gives us the state of the world (robot positions, etc.) we also need to listen for commands from the Referee Box.

The refbox from previous years can be found here: https://github.com/RoboCup-SSL/ssl-refbox
There had been talk about re-writing it for RoboCup 2019. The new repository can be found here: https://github.com/RoboCup-SSL/ssl-game-controller

There is sample client code in the RefBox repositories. Our code can likely be based very closely on that, and will likely be very similar to our code that receives SSL-Vision data (see the SSLVisionClient in the `backend_input` node)

Data that is received should be stored in the datastructure created in #43 , so this ticket is partially blocked by #43 . Nearly all the work can be done without the datastructure however.

### PR DESCRIPTION
**Pull request by jordanzliu**
### Description
Read the game state from the protobuf messages emitted by SSL Game Controller (Refbox replacement) and provide it to the AI Game State class.

### Testing Done

Tested with SSL Game Controller. (No unit tests are there for any other network_input classes)

### Resolved Issues

Resolves #72 

### Length Justification

A large chunk of this PR is a protobuf message copypasted from ssl-game-controller

### Review Checklist
*(Please check every item to indicate your code complies with it (by changing `[ ]`->`[x]`). This will hopefully save both you and the reviewer(s) a lot of time!)*
***It is the reviewers responsibility to also make sure every item here has been covered***
- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom` 
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

***Feel free to make additions of things that we should be checking to this file if you think there's something missing!!!!***

### CODE REVIEW

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp
Comment (garethellis0) [inline src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp]: I believe #33 is resolved now, so we can do this? See `ssl_vision_client.cp` for an example.

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h
Comment (garethellis0) [inline src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h:50]: Could we add a quick comment for each member function?

File: src/thunderbots/software/proto/src/ssl_referee.proto
Comment (garethellis0) [inline src/thunderbots/software/proto/src/ssl_referee.proto]: nit; github complaining about end-of-file

File: src/thunderbots/software/util/ros_messages.cpp
Comment (garethellis0) [inline src/thunderbots/software/util/ros_messages.cpp]: nit; github complaining about end-of-file

**Review by garethellis0 — CHANGES_REQUESTED** [review]
This is looking good! Nice work.

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp:24]: Not sure if that's necessary? I'm assuming you want the handling to be logging a proper error message and re-throwing the exception? Giving an outright invalid IP address is a pretty significant failure we'd want to take down the system right away, so I think it's fine to not catch an exception here.

File: src/thunderbots/software/util/ros_messages.cpp
Comment (MathewMacDougall) [inline src/thunderbots/software/util/ros_messages.cpp]: Is this including the same thing as the `#include` above? In any case, this should use the absolute path not the relative path.

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Just a few small comments but otherwise looks great!

File: src/thunderbots/software/network_input/backend.cpp
Comment (jordanzliu) [reply src/thunderbots/software/network_input/backend.cpp]: done

File: src/thunderbots/software/network_input/backend.h
Comment (jordanzliu) [reply src/thunderbots/software/network_input/backend.h:117]: done

File: src/thunderbots/software/network_input/main.cpp
Comment (jordanzliu) [reply src/thunderbots/software/network_input/main.cpp:84]: I'll open an issue for this

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp
Comment (jordanzliu) [reply src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp]: added error logging to ssl_gamecontroller_client

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h
Comment (jordanzliu) [reply src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h:50]: i fixed the existing javadoc function comments

File: src/thunderbots/software/proto/src/ssl_referee.proto
Comment (jordanzliu) [reply src/thunderbots/software/proto/src/ssl_referee.proto]: fixed

File: src/thunderbots/software/ai/ai.cpp
Comment (jordanzliu) [reply src/thunderbots/software/ai/ai.cpp]: fixed

File: src/thunderbots/software/util/ros_messages.cpp
Comment (jordanzliu) [reply src/thunderbots/software/util/ros_messages.cpp]: fixed the include

Comment (jordanzliu) [reply src/thunderbots/software/util/ros_messages.cpp]: I think i fixed the thing with the exception too

File: src/thunderbots/software/network_input/main.cpp
Comment (garethellis0) [reply src/thunderbots/software/network_input/main.cpp:84]: Sounds good! Maybe just leave a quick comment on both instances in this file linking to the new issue.

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp
Comment (garethellis0) [reply src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.cpp:24]: Fair enough. We'll actually catch and log this exception at the top level anyhow (but then still fail).

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h
Comment (garethellis0) [reply src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h:50]: Sorry! I mis-typed, I meant for each member _variable_.

**Review by MathewMacDougall — COMMENTED** [review]
LGTM pending Gareth's last clarification about member variable comments.

**Comment by MathewMacDougall** [comment]
@jordanzliu looks like some conflicts appeared (not sure how)

**Comment by jordanzliu** [comment]
yeah idk what's going on after i merged in the latest changes from master, but it's not compiling for me now

**Comment by jordanzliu** [comment]
i messed up something with the protobuf messages, for some reason protobuf compile fails silently. 
EDIT: there's naming conflicts between some newly-updated protobuf files from SSL Game Controller
EDIT 2: something is seriously, seriously wrong with protobuf and now it's not linking to the compiled protobuf object files.

**Comment by MathewMacDougall** [comment]
That's super weird. Let me know if you want me to look at anything.

**Comment by garethellis0** [comment]
@jordanzliu - looks like travis kinda choked here. Could you push another commit to force a rebuild?

**Comment by jordanzliu** [comment]
Everything seems to work fine now

**Comment by garethellis0** [comment]
:+1: - looks like there are just a few small comments left, then this should be good to go!

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Sorry, 2 more really small things but otherwise LGTM.

File: src/thunderbots/software/network_input/backend.cpp
Comment (MathewMacDougall) [reply src/thunderbots/software/network_input/backend.cpp:240]: Ah ok. Nevermind then.

File: src/thunderbots/software/network_input/main.cpp
Comment (jordanzliu) [reply src/thunderbots/software/network_input/main.cpp:84]: https://github.com/UBC-Thunderbots/Software/issues/275 
issue opened, added to comments

File: src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h
Comment (jordanzliu) [reply src/thunderbots/software/network_input/networking/ssl_gamecontroller_client.h:50]: commented

File: src/thunderbots/software/util/refbox_constants.h
Comment (jordanzliu) [reply src/thunderbots/software/util/refbox_constants.h]: renamed

**Review by MathewMacDougall — DISMISSED** [review]
LGTM :+1: Awesome work! :tada:

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada: - nice job!

Looks like we'll need to update this branch though, probably will need to re-approve after.

**Review by MathewMacDougall — APPROVED** [review]

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Pull request merged**
